patches/server/1048-Add-paper-version-util-class.patch

```diff
@@ -0,0 +1,211 @@
++        return this.implementedAPIs.contains(api);
++    }
++    private static class ServerImplementationInfo {
++        private final Set<Key> implementedAPIs = new HashSet<>();
++        private String latestName;
++        private ServerImplementationInfo() {
++            this.add("minecraft", "Vanilla");
++            this.add("papermc", "Paper");
++        }
++        @SuppressWarnings("PatternValidation")
++        private void add(@KeyPattern.Namespace String namespace, String name) {
++            this.implementedAPIs.add(Key.key(namespace, name.toLowerCase(Locale.ROOT)));
++            this.latestName = name;
++        }
++    }
++    private interface VersionInfo {
++
++        boolean is(int @NotNull ... version);
++
++        boolean is(@NotNull String version);
++
++        boolean isAtLeast(int @NotNull ... version);
++    }
++    private static class StableVersionInfo implements VersionInfo {
++        private final String version = SharedConstants.getCurrentVersion().getId();
++        private final int[] versionArr = Stream.of(this.version.split("\\.")).mapToInt(Integer::parseInt).toArray();
++        @Override
++        public boolean is(final int @NotNull ... version) {
++            return Arrays.equals(this.versionArr, version);
++        }
++
++        @Override
++        public boolean is(@NotNull final String version) {
++            return Objects.equals(this.version, version);
++        }
++
++        @Override
++        public boolean isAtLeast(final int @NotNull ... version) {
++            int maxLen = Math.max(this.versionArr.length, version.length);
++            for (int i = 0; i < maxLen; i++) {
++                int our = i < this.versionArr.length ? this.versionArr[i] : 0;
++                int requested = i < version.length ? version[i] : 0;
++                if (our != requested) {
++                    return our > requested;
++                }
++            }
++            return true;
++        }
++    }
++    private static class NonStableVersionInfo implements VersionInfo {
++        private static final Logger LOGGER = LogUtils.getClassLogger();
++        private static final String manifestURL = Optional.ofNullable(System.getProperty("paper.api.manifest-url"))
++            .orElse("https://piston-meta.mojang.com/mc/game/version_manifest_v2.json");
++        private final String version = SharedConstants.getCurrentVersion().getName();
++        private final String id = SharedConstants.getCurrentVersion().getId();
++        private final List<String> versionList;
++        private NonStableVersionInfo() {
++            HttpClient client = HttpClient.newHttpClient();
++            HttpRequest request = HttpRequest.newBuilder()
++                .uri(URI.create(manifestURL))
++                .build();
++            List<String> versionList = new ArrayList<>();
++            try {
++                HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
```

Comment:
Http requests should not be done in a ctor as it is a relativly heavy operation. It should be extracted into a factory method or better generated during build time as the currently running unstable version can only support version < current unstable version that has been built.


So a check of isAtleast("1.20.4") against 24w09a would succeed cause 1.20.4 is in the json array.
isAtLeast("1.20.5") doesn't succeed against 24w09a as the value is not contained in the built json array.
It basically doesn't make sense to call the endpoint on every startup if the minecraft version doesn't change without replacing the jar.

Reply:
As I said before, it's mostly to give idea how it will work on snapshots, i don't want to go deeper and change the build script of the server to work with this, and don't think that paper supporting snapshots will come soon either

And even if it does, snapshot version will most probably be unstable anyways, and mostly for devs to start updating their plugins, a snapshot paper shouldn't ever be on production, so this code won't be on a production server

Also the idea of downloading the file on build time has a big flaw: Paper may not want to jump to next snapshot/pre release right away, but decide to do once a few releases, so building a bit later the paper, after a new version came out, will break this class

Reply:
When paper will support snapshots and prereleases, only then it will really possible to look at how this could be made properly, but now it's too unknown how that snapshot support from paper will even look like to be able to think how can this be better

Reply:
Then at least move the data fetching into a static factory method, instantiate the unstable versioninfo using the fetched data and return that.

Reply:
> Then at least move the data fetching into a static factory method, instantiate the unstable versioninfo using the fetched data and return that.

Could you show a pseudo code of what you mean? I just don't understand the difference from what it's now, as we still will need to block the object initialization to get the version data. But I highly suspect that I just don't think correctly of what you mean with that change.

Reply:
```suggestion
diff --git a/src/main/java/io/papermc/paper/util/misc/ServerInfoProviderImpl.java b/src/main/java/io/papermc/paper/util/misc/ServerInfoProviderImpl.java
index 1364afc507..bacc415fbe 100644
--- a/src/main/java/io/papermc/paper/util/misc/ServerInfoProviderImpl.java
+++ b/src/main/java/io/papermc/paper/util/misc/ServerInfoProviderImpl.java
@@ -43,7 +43,7 @@ public class ServerInfoProviderImpl implements ServerInfoProvider {
         } else {
             // Snapshot, pre-release, etc
             // Supported the best we can
-            this.version = new NonStableVersionInfo();
+            this.version = NonStableVersionInfo.create();
         }
     }
     @Override
@@ -147,7 +147,12 @@ public class ServerInfoProviderImpl implements ServerInfoProvider {
         private final String version = SharedConstants.getCurrentVersion().getName();
         private final String id = SharedConstants.getCurrentVersion().getId();
         private final List<String> versionList;
-        private NonStableVersionInfo() {
+
+        private NonStableVersionInfo(List<String> versionList) {
+            this.versionList = ImmutableList.copyOf(versionList);
+        }
+
+        private static NonStableVersionInfo create() {
             HttpClient client = HttpClient.newHttpClient();
             HttpRequest request = HttpRequest.newBuilder()
                 .uri(URI.create(manifestURL))
@@ -162,8 +167,9 @@ public class ServerInfoProviderImpl implements ServerInfoProvider {
                 }
             } catch (IOException | InterruptedException e) {
                 LOGGER.error("Failed to fetch version manifest, plugins depending on version check may not work correctly", e);
+                return null;
             }
-            this.versionList = ImmutableList.copyOf(versionList);
+            return new NonStableVersionInfo(versionList);
         }
         @Override
         public boolean is(final int @NotNull ... version) {

```